lgc/patch/PatchPeepholeOpt.cpp

```diff
@@ -555,8 +555,9 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
     }
   }
 
-  // Do not clone allocas -- we don't want to potentially introduce them in the middle of the function.
-  if (prevIncomingInst && !isa<AllocaInst>(prevIncomingInst)) {
+  // Do not clone allocas, atomics and instructions with side effects.
+  if (prevIncomingInst && !isa<AllocaInst>(prevIncomingInst) && !prevIncomingInst->isAtomic() &&
+      !prevIncomingInst->mayHaveSideEffects()) {
```

Comment:
Isn't an atomic marked as having side effects?

Reply:
isAtomic is not a subset of mayHaveSideEffects.  Basically hasSideEffects is based on mayWriteToMemory, and isAtomic is based on the ordering constraints of the instruction.  My opinion is we definitely need mayHaveSideEffects, the isAtomic is basically an attempt at forward conservatism -- as I can't see a good reason to allow duplication of atomics which don't have side effects (e.g. atomic reads of some kind).

Reply:
Fair enough.